src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,470 @@
+        if ((*sled)[sfFlags] & lsfDisallowXRP)
+            return tecNO_TARGET;
+    }
+
+    // Create PayChan in ledger
+    auto const slep = std::make_shared<SLE> (
+        keylet::payChan (account, dst, (*sle)[sfSequence] - 1));
+    // Funds held in this channel
+    (*slep)[sfAmount] = ctx_.tx[sfAmount];
+    // Amount channel has already paid
+    (*slep)[sfBalance] = ctx_.tx[sfAmount].zeroed ();
+    (*slep)[sfAccount] = account;
+    (*slep)[sfDestination] = dst;
+    (*slep)[sfSettleDelay] = ctx_.tx[sfSettleDelay];
+    (*slep)[sfPublicKey] = ctx_.tx[sfPublicKey];
+    (*slep)[~sfCancelAfter] = ctx_.tx[~sfCancelAfter];
+    (*slep)[~sfSourceTag] = ctx_.tx[~sfSourceTag];
+    (*slep)[~sfDestinationTag] = ctx_.tx[~sfDestinationTag];
+
+    ctx_.view ().insert (slep);
+
+    // Add PayChan to owner directory
+    {
+        uint64_t page;
+        auto result = dirAdd (ctx_.view (), page, keylet::ownerDir (account),
+            slep->key (), describeOwnerDir (account),
+            ctx_.app.journal ("View"));
+        if (!isTesSuccess (result.first))
+            return result.first;
+        (*slep)[sfOwnerNode] = page;
+    }
+
+    // Deduct owner's balance, increment owner count
+    (*sle)[sfBalance] = (*sle)[sfBalance] - ctx_.tx[sfAmount];
+    (*sle)[sfOwnerCount] = (*sle)[sfOwnerCount] + 1;
+    ctx_.view ().update (sle);
+
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanFund::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    if (!isXRP (ctx.tx[sfAmount]) || (ctx.tx[sfAmount] <= beast::zero))
+        return temBAD_AMOUNT;
+
+    return preflight2 (ctx);
+}
+
+TER
+PayChanFund::doApply()
+{
+    Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
+    auto const slep = ctx_.view ().peek (k);
+    if (!slep)
+        return tecNO_ENTRY;
+
+    AccountID const src = (*slep)[sfAccount];
+    auto const txAccount = ctx_.tx[sfAccount];
+    auto const expiration = (*slep)[~sfExpiration];
+
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (expiration && closeTime >= *expiration))
+            return closeChannel (
+                slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
+    }
+
+    if (src != txAccount)
+        // only the owner can add funds or extend
+        return tecNO_PERMISSION;
+
+    if (auto extend = ctx_.tx[~sfExpiration])
+    {
+        auto minExpiration =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count () +
+                (*slep)[sfSettleDelay];
+        if (expiration && *expiration < minExpiration)
```

Comment:
Yeah, I knew about that (I know because Scott Schurr used it in some code I reviewed). I prefer the non-magic version.